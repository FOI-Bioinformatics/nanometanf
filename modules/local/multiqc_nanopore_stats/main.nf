process MULTIQC_NANOPORE_STATS {
    tag "$prefix"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/python:3.11' :
        'biocontainers/python:3.11' }"

    input:
    val(sample_stats)  // List of sample statistics from various sources
    val(prefix)

    output:
    path "*_mqc.json"          , emit: multiqc_files
    path "*_mqc.yaml"          , emit: multiqc_yaml
    path "versions.yml"        , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    """
    #!/usr/bin/env python3
    import json
    import yaml
    import sys
    from collections import defaultdict

    sample_stats = ${groovy.json.JsonOutput.toJson(sample_stats)}

    # Generate MultiQC custom content for nanopore statistics
    # Format: https://multiqc.info/docs/#custom-content

    # 1. General Statistics Table
    general_stats = {
        "id": "nanopore_general_stats",
        "section_name": "Nanopore Sequencing Statistics",
        "description": "Overview of Oxford Nanopore sequencing run metrics",
        "plot_type": "table",
        "pconfig": {
            "id": "nanopore_stats_table",
            "title": "Nanopore Run Summary"
        },
        "data": {}
    }

    # 2. Read Length Distribution
    read_length_data = {
        "id": "nanopore_read_length",
        "section_name": "Read Length Distribution",
        "description": "Distribution of read lengths across all samples",
        "plot_type": "linegraph",
        "pconfig": {
            "id": "nanopore_read_length_plot",
            "title": "Read Length Distribution",
            "xlab": "Read Length (bp)",
            "ylab": "Number of Reads",
            "yLog": True
        },
        "data": {}
    }

    # 3. Quality Score Distribution
    quality_data = {
        "id": "nanopore_quality",
        "section_name": "Quality Score Distribution",
        "description": "Mean quality scores across samples",
        "plot_type": "bargraph",
        "pconfig": {
            "id": "nanopore_quality_plot",
            "title": "Quality Score Distribution",
            "ylab": "Mean Quality Score"
        },
        "data": {}
    }

    # Parse sample statistics
    for sample in sample_stats:
        sample_id = sample.get('sample_id', 'unknown')

        # General stats
        general_stats['data'][sample_id] = {
            'Total Reads': sample.get('total_reads', 0),
            'Total Bases': sample.get('total_bases', 0),
            'Mean Read Length': sample.get('mean_length', 0),
            'Median Read Length': sample.get('median_length', 0),
            'N50': sample.get('n50', 0),
            'Mean Quality': sample.get('mean_quality', 0)
        }

        # Read length distribution (if available)
        if 'read_length_histogram' in sample:
            read_length_data['data'][sample_id] = sample['read_length_histogram']

        # Quality distribution
        if 'mean_quality' in sample:
            quality_data['data'][sample_id] = {
                'Quality Score': sample['mean_quality']
            }

    # Write MultiQC JSON files
    with open('${prefix}_nanopore_stats_mqc.json', 'w') as f:
        json.dump(general_stats, f, indent=2)

    with open('${prefix}_read_length_mqc.json', 'w') as f:
        json.dump(read_length_data, f, indent=2)

    with open('${prefix}_quality_mqc.json', 'w') as f:
        json.dump(quality_data, f, indent=2)

    # Write YAML config for MultiQC
    multiqc_config = {
        'custom_data': {
            'nanopore_stats': {
                'file_format': 'json',
                'section_name': 'Nanopore Sequencing',
                'plot_type': 'table'
            }
        }
    }

    with open('${prefix}_multiqc_config_mqc.yaml', 'w') as f:
        yaml.dump(multiqc_config, f)

    print(f"Generated MultiQC custom content for {len(sample_stats)} samples", file=sys.stderr)

    # Write versions
    with open('versions.yml', 'w') as f:
        f.write('"${task.process}":\\n')
        f.write(f'  python: {sys.version.split()[0]}\\n')
    """

    stub:
    """
    touch ${prefix}_nanopore_stats_mqc.json
    touch ${prefix}_read_length_mqc.json
    touch ${prefix}_quality_mqc.json
    touch ${prefix}_multiqc_config_mqc.yaml

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 --version | sed 's/Python //g')
    END_VERSIONS
    """
}
