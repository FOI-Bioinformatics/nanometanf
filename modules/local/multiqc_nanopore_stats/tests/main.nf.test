nextflow_process {

    name "Test Process MULTIQC_NANOPORE_STATS"
    script "../main.nf"
    process "MULTIQC_NANOPORE_STATS"

    tag "modules"
    tag "modules_local"
    tag "multiqc_nanopore_stats"

    test("multiqc_nanopore_stats - basic functionality") {

        when {
            process {
                """
                input[0] = [
                    [
                        sample_id: 'sample1',
                        total_reads: 10000,
                        total_bases: 50000000,
                        mean_length: 5000,
                        median_length: 4800,
                        n50: 6500,
                        mean_quality: 12.5
                    ],
                    [
                        sample_id: 'sample2',
                        total_reads: 15000,
                        total_bases: 75000000,
                        mean_length: 5000,
                        median_length: 4900,
                        n50: 7000,
                        mean_quality: 13.2
                    ]
                ]
                input[1] = 'test_nanopore'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("multiqc_nanopore_stats - single sample") {

        when {
            process {
                """
                input[0] = [
                    [
                        sample_id: 'single_sample',
                        total_reads: 5000,
                        total_bases: 25000000,
                        mean_length: 5000,
                        median_length: 4700,
                        n50: 6000,
                        mean_quality: 11.8
                    ]
                ]
                input[1] = 'single_test'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("multiqc_nanopore_stats - with histogram data") {

        when {
            process {
                """
                input[0] = [
                    [
                        sample_id: 'histogram_sample',
                        total_reads: 8000,
                        total_bases: 40000000,
                        mean_length: 5000,
                        median_length: 4850,
                        n50: 6200,
                        mean_quality: 12.1,
                        read_length_histogram: [
                            1000: 100,
                            2000: 250,
                            3000: 400,
                            4000: 500,
                            5000: 600,
                            6000: 450,
                            7000: 300,
                            8000: 150
                        ]
                    ]
                ]
                input[1] = 'histogram_test'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("multiqc_nanopore_stats - stub run") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [
                        sample_id: 'stub_sample',
                        total_reads: 1000,
                        total_bases: 5000000,
                        mean_length: 5000,
                        median_length: 4800,
                        n50: 6000,
                        mean_quality: 12.0
                    ]
                ]
                input[1] = 'stub_test'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("multiqc_nanopore_stats - empty stats list") {

        when {
            process {
                """
                input[0] = []
                input[1] = 'empty_test'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("multiqc_nanopore_stats - high quality reads") {

        when {
            process {
                """
                input[0] = [
                    [
                        sample_id: 'high_quality',
                        total_reads: 12000,
                        total_bases: 60000000,
                        mean_length: 5000,
                        median_length: 4900,
                        n50: 7500,
                        mean_quality: 15.5
                    ]
                ]
                input[1] = 'high_quality_test'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
