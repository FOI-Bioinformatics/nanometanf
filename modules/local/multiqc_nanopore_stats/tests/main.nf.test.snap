{
    "multiqc_nanopore_stats - basic functionality": {
        "content": [
            {
                "0": [
                    [
                        "test_nanopore_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test_nanopore_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test_nanopore_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "test_nanopore_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ],
                "multiqc_files": [
                    [
                        "test_nanopore_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test_nanopore_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test_nanopore_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "multiqc_yaml": [
                    "test_nanopore_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:32:17.475619"
    },
    "multiqc_nanopore_stats - with histogram data": {
        "content": [
            {
                "0": [
                    [
                        "histogram_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "histogram_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "histogram_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "histogram_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ],
                "multiqc_files": [
                    [
                        "histogram_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "histogram_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "histogram_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "multiqc_yaml": [
                    "histogram_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:32:25.445554"
    },
    "multiqc_nanopore_stats - stub run": {
        "content": [
            {
                "0": [
                    [
                        "stub_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "stub_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "stub_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "stub_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ],
                "multiqc_files": [
                    [
                        "stub_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "stub_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "stub_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "multiqc_yaml": [
                    "stub_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:32:29.301452"
    },
    "multiqc_nanopore_stats - empty stats list": {
        "content": [
            {
                "0": [
                    [
                        "empty_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "empty_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "empty_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "empty_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ],
                "multiqc_files": [
                    [
                        "empty_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "empty_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "empty_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "multiqc_yaml": [
                    "empty_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:32:33.361646"
    },
    "multiqc_nanopore_stats - high quality reads": {
        "content": [
            {
                "0": [
                    [
                        "high_quality_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "high_quality_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "high_quality_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "high_quality_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ],
                "multiqc_files": [
                    [
                        "high_quality_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "high_quality_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "high_quality_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "multiqc_yaml": [
                    "high_quality_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:32:37.419241"
    },
    "multiqc_nanopore_stats - single sample": {
        "content": [
            {
                "0": [
                    [
                        "single_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "single_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "single_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "single_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ],
                "multiqc_files": [
                    [
                        "single_test_nanopore_stats_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "single_test_quality_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "single_test_read_length_mqc.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "multiqc_yaml": [
                    "single_test_multiqc_config_mqc.yaml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,778ee7e467c71f4062c2689367ff8200"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:32:21.367387"
    }
}